form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkTaskProvider.kt

```diff
@@ -16,25 +16,106 @@
+import org.apache.logging.log4j.util.Strings
+import org.camunda.bpm.engine.RepositoryService
+import org.camunda.bpm.engine.RuntimeService
 import org.camunda.bpm.engine.task.Task
 
 class FormFlowProcessLinkTaskProvider(
-    val formFlowService: FormFlowService
+    private val formFlowService: FormFlowService,
+    private val formAssociationService: FormAssociationService,
+    private val documentService: DocumentService,
+    private val repositoryService: RepositoryService,
+    private val runtimeService: RuntimeService,
 ): ProcessLinkTaskProvider<FormFlowTaskOpenResultProperties> {
-    private val FORM_FLOW_TASK_TYPE_KEY = "form-flow"
 
     override fun supports(formLink: FormLink?): Boolean {
         return formLink is BpmnElementFormFlowIdLink
     }
 
     override fun getTaskResult(task: Task, formLink: FormLink): TaskOpenResult<FormFlowTaskOpenResultProperties> {
         val instances = formFlowService.findInstances(mapOf("taskInstanceId" to task.id))
-        assert(instances.size == 1)
-        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instances[0].id.id))
+        val instance = when (instances.size) {
+            0 -> createFormFlowInstance(task)
+            1 -> instances[0]
+            else -> throw IllegalStateException("Multiple form flow instances linked to task: ${task.id}")
+        }
+        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instance.id.id))
     }
-}
\ No newline at end of file
+
+    private fun createFormFlowInstance(task: Task): FormFlowInstance {
+        val additionalProperties: MutableMap<String, Any> = getAdditionalProperties(task)
+        val formFlowIdAsArray = getFormAssociationByTask(task).formLink.formFlowId.split(":")
+        val formFlowDefinition: FormFlowDefinition = getFormFlowDefinition(formFlowIdAsArray)
+        return formFlowService.save(formFlowDefinition.createInstance(additionalProperties))
+    }
+
+    private fun getFormAssociationByTask(task: Task): FormAssociation {
+        val processDefinition = repositoryService.createProcessDefinitionQuery()
+            .processDefinitionId(task.processDefinitionId)
+            .singleResult()
+
+        val formAssociation = formAssociationService.getFormAssociationByFormLinkId(
+            processDefinition.key,
+            task.taskDefinitionKey
+        ).orElseThrow { IllegalStateException("No form association found. Process: '${processDefinition.key}', task: '${task.taskDefinitionKey}'") }
+
+        if (formAssociation.formLink !is BpmnElementFormFlowIdLink) {
+            throw IllegalStateException("Found form association is not of type 'BpmnElementFormFlowIdLink'. Type: ${formAssociation.formLink.javaClass.simpleName}, process: '${processDefinition.key}', task: '${task.taskDefinitionKey}'")
+        }
+
+        return formAssociation
+    }
+
+    private fun getFormFlowDefinition(formFlowIdAsArray: List<String>): FormFlowDefinition {
```

Comment:
i don't like this very much. the knowledge that the id is a 2 part thing is now split over 2 methods. Why not pass in the whole id and split it in this method.

Reply:
fixed